embassy-stm32/src/dma/bdma.rs

```diff
@@ -257,3 +253,162 @@ pac::interrupts! {
+        dma: pac::bdma::Dma,
+        channel_number: u8,
+        state_number: u8,
+        request: Request,
+        dir: vals::Dir,
+        peri_addr: *const u32,
+        mem_addr: *mut u32,
+        mem_len: usize,
+        incr_mem: bool,
+        data_size: vals::Size,
+        #[cfg(dmamux)] dmamux_regs: pac::dmamux::Dmamux,
+        #[cfg(dmamux)] dmamux_ch_num: u8,
+    ) -> impl Future<Output = ()> {
+        // ndtr is max 16 bits.
+        assert!(mem_len <= 0xFFFF);
+
+        // Reset status
+        reset_status(dma, channel_number);
+
+        let on_drop = OnDrop::new(move || unsafe {
+            stop(dma, channel_number);
+        });
+
+        low_level_api::start_transfer(
+            dma,
+            channel_number,
+            #[cfg(any(bdma_v2, dmamux))]
+            request,
+            dir,
+            peri_addr,
+            mem_addr,
+            mem_len,
+            incr_mem,
+            data_size,
+            #[cfg(dmamux)]
+            dmamux_regs,
+            #[cfg(dmamux)]
+            dmamux_ch_num,
+        );
+
+        async move {
+            let res = low_level_api::wait_for_completion(dma, state_number, channel_number).await;
+
+            drop(on_drop)
+        }
+    }
+
+    pub unsafe fn start_transfer(
+        dma: pac::bdma::Dma,
+        channel_number: u8,
+        #[cfg(any(bdma_v2, dmamux))] request: Request,
+        dir: vals::Dir,
+        peri_addr: *const u32,
+        mem_addr: *mut u32,
+        mem_len: usize,
+        incr_mem: bool,
+        data_size: vals::Size,
+        #[cfg(dmamux)] dmamux_regs: pac::dmamux::Dmamux,
+        #[cfg(dmamux)] dmamux_ch_num: u8,
+    ) {
+        let ch = dma.ch(channel_number as _);
+
+        #[cfg(dmamux)]
+        super::super::dmamux::configure_dmamux(dmamux_regs, dmamux_ch_num, request);
+
+        #[cfg(bdma_v2)]
+        critical_section::with(|_| {
+            dma.cselr()
+                .modify(|w| w.set_cs(channel_number as _, request))
+        });
+
+        // "Preceding reads and writes cannot be moved past subsequent writes."
+        fence(Ordering::Release);
+
+        ch.par().write_value(peri_addr as u32);
+        ch.mar().write_value(mem_addr as u32);
+        ch.ndtr().write(|w| w.set_ndt(mem_len as u16));
+        ch.cr().write(|w| {
+            w.set_psize(data_size);
+            w.set_msize(data_size);
+            if incr_mem {
+                w.set_minc(vals::Inc::ENABLED);
+            } else {
+                w.set_minc(vals::Inc::DISABLED);
+            }
+            w.set_dir(dir);
+            w.set_teie(true);
+            w.set_tcie(true);
+            w.set_en(true);
+        });
+    }
+
+    pub unsafe fn stop(dma: pac::bdma::Dma, ch: u8) {
+        let ch = dma.ch(ch as _);
+
+        // Disable the channel and interrupts with the default value.
+        ch.cr().write(|_| ());
+
+        // Wait for the transfer to complete when it was ongoing.
+        while ch.cr().read().en() {}
```

Comment:
Instead of spinning, it might be better to simply make `stop` request the stop, but not wait until actually stopped. (documenting it in the Channel trait method, and maybe renaming `stop` to `request_stop` or something else?)

The advantages are:

- User code decides whether to spin or not (for example using async instead. Stopping is probably very fast though, so using async is probably not worth it)
- It's possible to do stuff like stopping 2 channels concurrently: request both to stop, and *then* spin waiting for both to stop.